apps/files_sharing/src/components/SharePermissionsEditor.vue

```diff
@@ -0,0 +1,291 @@
+					:disabled="saving"
+					:name="randomFormName"
+					class="sharing-entry__action--public-upload"
+					@change="setSharePermissions(bundledPermissions.FILE_DROP)">
+					{{ t('files_sharing', 'File drop (upload only)') }}
+				</ActionRadio>
+
+				<!-- custom permissions button -->
+				<ActionButton :title="t('files_sharing', 'Custom permissions')"
+					@click="showCustomPermissionsForm = true">
+					<template #icon>
+						<Tune />
+					</template>
+					{{ sharePermissionsIsBundle ? "" : sharePermissionsSummary }}
+				</ActionButton>
+			</template>
+
+			<!-- custom permissions -->
+			<span v-else :class="{error: !sharePermissionsSetIsValid}">
+				<ActionCheckbox :checked="shareHasPermissions(atomicPermissions.READ)"
+					:disabled="saving || !canToggleSharePermissions(atomicPermissions.READ)"
+					@update:checked="toggleSharePermissions(atomicPermissions.READ)">
+					{{ t('files_sharing', 'Read') }}
+				</ActionCheckbox>
+				<ActionCheckbox :checked="shareHasPermissions(atomicPermissions.CREATE)"
+					:disabled="saving || !canToggleSharePermissions(atomicPermissions.CREATE)"
+					@update:checked="toggleSharePermissions(atomicPermissions.CREATE)">
+					{{ t('files_sharing', 'Create') }}
+				</ActionCheckbox>
+				<ActionCheckbox :checked="shareHasPermissions(atomicPermissions.UPDATE)"
+					:disabled="saving || !canToggleSharePermissions(atomicPermissions.UPDATE)"
+					@update:checked="toggleSharePermissions(atomicPermissions.UPDATE)">
+					{{ t('files_sharing', 'Update') }}
+				</ActionCheckbox>
+				<ActionCheckbox :checked="shareHasPermissions(atomicPermissions.DELETE)"
+					:disabled="saving || !canToggleSharePermissions(atomicPermissions.DELETE)"
+					@update:checked="toggleSharePermissions(atomicPermissions.DELETE)">
+					{{ t('files_sharing', 'Delete') }}
+				</ActionCheckbox>
+
+				<ActionButton @click="showCustomPermissionsForm = false">
+					<template #icon>
+						<ChevronLeft />
+					</template>
+					{{ t('files_sharing', 'Bundle permissions') }}
+				</ActionButton>
+			</span>
+		</template>
+	</span>
+</template>
+
+<script>
+import ActionButton from '@nextcloud/vue/dist/Components/ActionButton'
+import ActionRadio from '@nextcloud/vue/dist/Components/ActionRadio'
+import ActionCheckbox from '@nextcloud/vue/dist/Components/ActionCheckbox'
+
+import SharesMixin from '../mixins/SharesMixin'
+import {
+	ATOMIC_PERMISSIONS,
+	BUNDLED_PERMISSIONS,
+	hasPermissions,
+	permissionsSetIsValid,
+	togglePermissions,
+	canTogglePermissions,
+} from '../lib/SharePermissionsToolBox'
+
+import Tune from 'vue-material-design-icons/Tune.vue'
+import ChevronLeft from 'vue-material-design-icons/ChevronLeft.vue'
+
+export default {
+	name: 'SharePermissionsEditor',
+
+	components: {
+		ActionButton,
+		ActionCheckbox,
+		ActionRadio,
+		Tune,
+		ChevronLeft,
+	},
+
+	mixins: [SharesMixin],
+
+	data() {
+		return {
+			randomFormName: Math.random().toString(27).substr(2),
```

Comment:
```suggestion
			randomFormName: Math.random().toString(27).substring(2),
```
[.substr()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr) has been deprecated

Reply:
Arf, sad that eslint is not picking that up, thanks :)